packages/nouns-contracts/contracts/governance/NounsDAOV3Proposals.sol

```diff
@@ -0,0 +1,685 @@
+        uint256[] values;
+        string[] signatures;
+        bytes[] calldatas;
+    }
+
+    /// @notice The maximum number of actions that can be included in a proposal
+    uint256 public constant proposalMaxOperations = 10; // 10 actions
+
+    bytes32 public constant DOMAIN_TYPEHASH =
+        keccak256('EIP712Domain(string name,uint256 chainId,address verifyingContract)');
+
+    bytes32 public constant PROPOSAL_TYPEHASH =
+        keccak256(
+            'Proposal(address proposer,address[] targets,uint256[] values,string[] signatures,bytes[] calldatas,string description,uint40 expiry)'
+        );
+
+    /**
+     * @notice Function used to propose a new proposal. Sender must have delegates above the proposal threshold
+     * @param txs Target addresses, eth values, function signatures and calldatas for proposal calls
+     * @param description String description of the proposal
+     * @return Proposal id of new proposal
+     */
+    function propose(
+        NounsDAOStorageV3.StorageV3 storage ds,
+        ProposalTxs memory txs,
+        string memory description
+    ) internal returns (uint256) {
+        ProposalTemp memory temp;
+        temp.totalSupply = ds.nouns.totalSupply();
+        temp.proposalThreshold = checkPropThreshold(ds, ds.nouns.getPriorVotes(msg.sender, block.number - 1));
+        checkProposaTxs(txs);
+        checkNoActiveProp(ds, msg.sender);
+
+        ds.proposalCount++;
+        NounsDAOStorageV3.Proposal storage newProposal = createNewProposal(
+            ds,
+            ds.proposalCount,
+            temp.proposalThreshold,
+            txs
+        );
+        ds.latestProposalIds[newProposal.proposer] = newProposal.id;
+
+        emitNewPropEvents(newProposal, new address[](0), ds.minQuorumVotes(), txs, description);
+
+        return newProposal.id;
+    }
+
+    function proposeBySigs(
+        NounsDAOStorageV3.StorageV3 storage ds,
+        NounsDAOStorageV3.ProposerSignature[] memory proposerSignatures,
+        ProposalTxs memory txs,
+        string memory description
+    ) internal returns (uint256) {
+        checkProposaTxs(txs);
+        uint256 proposalId = ds.proposalCount = ds.proposalCount + 1;
+
+        bytes memory proposalEncodeData = calcProposalEncodeData(txs, description);
+
+        uint256 votes;
+        address[] memory signers = new address[](proposerSignatures.length);
+        for (uint256 i = 0; i < proposerSignatures.length; ++i) {
+            verifyProposalSignature(ds, proposalEncodeData, proposerSignatures[i]);
+            address signer = signers[i] = proposerSignatures[i].signer;
+
+            checkNoActiveProp(ds, signer);
+            ds.latestProposalIds[signer] = proposalId;
+            votes += ds.nouns.getPriorVotes(signer, block.number - 1);
+        }
+
+        checkNoActiveProp(ds, msg.sender);
+        ds.latestProposalIds[msg.sender] = proposalId;
+        votes += ds.nouns.getPriorVotes(msg.sender, block.number - 1);
```

Comment:
@eladmallel `proposal.proposer` votes should also be counted `cancel`

Reply:
fixed!